Added plugin tangoinfo

ix5 opened the pull request:
See the [old one](https://github.com/metabrainz/picard-plugins/pull/49), can't reopen so I created a new one.
All works now.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+try:
+    from urllib.request import urlopen as urlopen
+except ImportError:
+    from urllib import urlopen
```
These imports are not used anywhere.

ix5 replied:
Fixed

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+
+        def __getitem__(self, name):
+            self.lock_for_read()
+            value = self.queue[name] if name in self.queue else None
```
It doesn't matter much in this case, but that's exactly what `value = self.queue.get(name)` would have done

ix5 replied:
Wouldn't it throw a keyerror if name wasn't in self.queue?

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+                    .lstrip("0")
+                    # For discnumbers like 02/12"
+        else:
+            discnumber = "1"
```
`discnr` at one point and `discnumber` at another can't be correct.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+            discnr = str(track_metadata["discnumber"])\
+                    .split("/")[0]\
+                    .lstrip("0")
+                    # For discnumbers like 02/12"
```
I think the discnumber tag in Picard will always be just an integer and not contain the number of total discs.

Sophist-UK replied:
See similar comment below - see also `totaldiscs`.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+        tracknr = str(track_metadata["tracknumber"])\
+                .split("/")[0]\
+                .lstrip("0")
+                # For tracknumbers like 02/12"
```
I think this will never happen.

Sophist-UK replied:
@mineo is correct - in Picard the values are already stored separately - see e.g. `totaltracks`. They are combined only when saved to some file formats and split when you load metadata from these same formats.

ix5 replied:
Will picard always return a discnumber? Or could the field also be empty? I have only added a check if discnumber exists, else it gets "1", because I need to construct the TINT.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+
+
+    def website_add_track(self, album, track, barcode):
+        #: returns the tango.info website, raw
```
It doesn't.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+
+    def website_add_track(self, album, track, barcode):
+        #: returns the tango.info website, raw
+        self.album_add_request(album) # increase counter, gets decreased later
```
It doesn't get descreased if the condition on the next line is False because `website_process` will never be called.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+                            .lstrip("0")
+                            # For discnumbers like 02/12"
+                else:
+                    discnumber = "1"
```
Again, `discnumber` vs. `discnr`.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+        if error:
+            log.error("%s: %s: Network error retrieving info for barcode", PLUGIN_NAME, barcode)
+            tuples = self.albumpage_queue.remove(barcode)
+            for track, album in tuples:
```
This code (and the same one in lines 158 & 159) looks suspicious to me: `album._requests` is incremented exactly once per album (/barcode), but is decremented for every track on the album, which is going to be more than once in most cases.

Sophist-UK replied:
This code is based on my Artist Official Website plugin - so I understand how it works.

Because this is a Track Metadata processor, album._requests is incremented once per track. To avoid duplicated requests for the same metadata, the tracks are queued against each request and when the request is completed, it is decremented once per track.

mineo replied:
Oh, now I see how it works. Seems a bit weird to count the number of not-really-started requests, but as long as the other direction is aware of it...

Sophist-UK replied:
It's complicated. An artist website request can be for some but not all tracks on an album (e.g. Various Artists), and for tracks across multiple albums. So its easier to increment and decrement once per track than to work out whether you need to increment / decrement because e.g. this is the first track on the album etc.

ix5 replied:
Yes, I just copied it. Maybe it does not make as much sense as for official websites, because each album can only have on tango.info page.

Sophist-UK replied:
I suppose you could simply store the request against an album and check if the request has already been queued or cached and then apply the response to all tracks in the album - but this might actually be more effort than it is worth.

ix5 replied:
And you'd still have to use a track processor and not an album one(individual genres, dates for each track etc.), so no real benefit there I guess. Also, it just occurred to me that there could be edge cases where MB and tango.info disagree, e.g. multi-volume compilations or something similar.
So your approach to queueing is the most sensible after all.
What irritates me is that if MB and tango.info both have new data(e.g date was 1950, MB has 2004 and tango.info 1950-01-01) the MB one takes preference.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+
+    def album_remove_request(self, album):
+        album._requests -= 1
+        if album._requests == 0:
```
`_finalize_loading` already performs this test.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+        if response.splitlines()[1] == "<title>Contents - tango.info</title>":
+            log.debug("%s: No album with barcode %s on tango.info" % (PLUGIN_NAME, barcode))
+            return None
+        else:
```
No need to put this into an else branch and the additional indentation level of that, the line before is a `return`.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+            return None
+        else:
+            regexp = """<h2><a href="\/tracks">Tracks<\/a><\/h2>(?!<\/table>)(.+?)<\/table>"""
+            trs = re.compile("<tr>((?!<\/tr>).+?)</tr>")
```
At the moment, these are recompiled on each call to `barcode_process_metadata` (ok, that's a lie, the `re` module caches them, but this still means a cache lookup somewhere). If you make these global constant variables, compiling would only happen once and they wouldn't be buried in this function.

ix5 replied:
Fixed, I put them outside the whole class, hope that is right.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+        tuples = self.albumpage_queue.remove(barcode)
+        #log.debug("%s: %r: tango.info page = http://tango.ingo/0%s", PLUGIN_NAME, barcode, barcode)
+        for track, album in tuples:
+            self.album_remove_request(album)
```
This means that `album._finalize_loading` will be called before the data from tango.info has been written on all tracks.

Sophist-UK replied:
Ah yes - that is probably also an issue in Artist Website plugin I guess.

ix5 replied:
Also fixed that in my plugin.

mineo commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+<p>Uses web scraping since opendata is disabled for now</p>'
+PLUGIN_VERSION = '0.1.3'
+PLUGIN_API_VERSIONS = ["0.15.0", "0.15.1", "0.16.0", "1.0.0", "1.1.0", "1.2.0", "1.3.0"]
+PLUGIN_LICENSE = 'ISC License'
```
Is relicensing GPL code under the ISC license allowed?

ix5 replied:
I had previously chosen ISC for my rough draft, but on the other PR @Sophist-UK told me to take inspiration from his `albumartist_website` plugin. Changed to GPLv2 since now it's mostly his code.

Sophist-UK commented on plugins/tangoinfo/tangoinfo.py:
```diff
@@ -0,0 +1,245 @@
+            log.debug("%s: No album with barcode %s on tango.info" % (PLUGIN_NAME, barcode))
+            return None
+        else:
+            regexp = """<h2><a href="\/tracks">Tracks<\/a><\/h2>(?!<\/table>)(.+?)<\/table>"""
```
You should also compile `regexp` and not used it in a raw `re` call.

ix5 replied:
Done. Does it have that much of an impact on performance though? Would be interesting to know for future projects.

ix5 commented:
Alright, thank you so much everyone. Updated version is here: https://github.com/ix5/picard-tangoinfo

ix5 commented:
Strings from tango.info are now converted to unicode, so no more unfinished loading and hanging because of names like *Félix* (that was a rookie mistake).

I have thoroughly tested the plugin with much of my collection. **It works very reliably**. 
A lot of times it seems MB and tango.info disagree on something. Nothing I can do about that, but I have added a notice in the debug log so users can rectify inaccuracies.

Feel free to test this on a Windows machine; if you need to, I can provide some sample songs(or download some Carlos Gardel tracks, copyright has expired on his works afaik).
